README.md
# Behavioral Cloning Project

[![Udacity - Self-Driving Car NanoDegree](https://s3.amazonaws.com/udacity-sdc/github/shield-carnd.svg)](http://www.udacity.com/drive)

# Overview

In this project, I used deep neural networks and convolutional neural networks to clone driving behavior. I trained, validated and tested a model using Keras. The model outputs a steering angle to an autonomous vehicle.

I used the provided simulator where I could steer a car around a track for data collection. I used image data and steering angles to train a neural network and then used this model to drive the car autonomously around the track.

# Dependencies

This lab requires:

* [CarND Term1 Starter Kit](https://github.com/udacity/CarND-Term1-Starter-Kit)

The lab enviroment can be created with CarND Term1 Starter Kit. Click [here](https://github.com/udacity/CarND-Term1-Starter-Kit/blob/master/README.md) for the details.

# The Project

The goals / steps of this project are the following:
* Use the simulator to collect data of good driving behavior 
* Design, train and validate a model that predicts a steering angle from image data
* Use the model to drive the vehicle autonomously around the first track in the simulator. The vehicle should remain on the road for an entire loop around the track.
* Summarize the results with a written report

## Rubric Points

### Files Submitted & Code Quality

#### 1. Submission includes all required files and can be used to run the simulator in autonomous mode

My project includes the following files:

* [model.py](model.py) containing the script to create and train the model
* [drive.py](drive.py) for driving the car in autonomous mode
* [model.h5](model.h5) containing a trained convolution neural network 
* [writeup_report.md](writeup_report.md) summarizing the results

#### 2. Submission includes functional code

Using the Udacity provided simulator and my drive.py file, the car can be driven autonomously around the track by executing 

```sh
python drive.py model.h5
```

#### 3. Submission code is usable and readable

The model.py file contains the code for training and saving the convolution neural network. The file shows the pipeline I used for training and validating the model, and it contains comments to explain how the code works. I did not implement data generator as the virtual machine I used for training had enough memory to load all the training data at once.

### Model Architecture and Training Strategy

#### 1. An appropriate model architecture has been employed

My model is based on the model proposed by NVIDIA developers in the [End-to-End Deep Learning for Self-Driving Cars](https://devblogs.nvidia.com/deep-learning-self-driving-cars/) article.

For mode details about the model see the `Final Model Architecture` section below.

#### 2. Attempts to reduce overfitting in the model

In order to reduce overfitting I've collected data from both track 1 and 2.
In addition, I applied data augmentation by adding a horizontally flipped version of each frame. Having so various data set was enough to successfully train the model. 

I split the collected data into trained and validated sets (as 80% and 20% respectively). The model was trained and validated using the data sets to ensure that the model was not overfitting. The model was tested by running it through the simulator and ensuring that the vehicle could stay on the track 1 as well as on the track 2.

#### 3. Model parameter tuning

The model used an adam optimizer, so the learning rate was not tuned manually (model.py line 120).

#### 4. Appropriate training data

Training data was chosen to keep the vehicle driving on the road. I used center lane driving without any recovering driving. However, I used images from the left and right cameras to train the model to recover. For more details see the next section. 

### Architecture and Training Documentation

#### 1. Solution Design Approach

My first step was to ensure the overall pipeline (capturing data, training the model using the data, and using the model to drive the car) is working.

I used the simplest neural network which consisted of one flatten layer and an output neuron. I recorded driving one lap on the track one, trained the NN, and made sure I'm able to run the simulator in autonomous mode.

Then I decided to use the LeNet-5 NN because I was already familiar with it and it showned pretty good results in the traffic sign classifier project.

In order to gauge how well the model was working, I split my image and steering angle data into a training and validation set. I did not face the overfitting problem, however the model did not work well: the car was'n able to pass the most curve parts of the track one, and it wasn't able to drive on the track two at all.

Next step I decided to try the CNN proposed by NVIDIA in the [End-to-End Deep Learning for Self-Driving Cars](https://devblogs.nvidia.com/deep-learning-self-driving-cars/) article. Using this CNN I acheived better performance (lower error) and it drove the car significantly better: it was able to keep the car within the track one as well as within the track two.

During training the models I ran the simulator to see how well the car was driving around track one and two. There were several issues which I solved by adding more training data. For more details see the section 3 below.

At the end of the process, the vehicle is able to drive autonomously around the tracks without leaving the road.

#### 2. Final Model Architecture

My final model is based on the model proposed by NVIDIA developers in the [End-to-End Deep Learning for Self-Driving Cars](https://devblogs.nvidia.com/deep-learning-self-driving-cars/) article.

The model consists of five convolutional layers followed by three fully-connected layers. Each layer is described in the table below:

|Layer | Output Shape | Configuration |
| ---  | ------------ | ------------- |
|Input           | 65x320x3  | Normalized input |
|Convolutional   | 31x158x24 | filters=24, kernel=5x5, strides=2x2 |
|Activation      | 31x158x24 | RELU |
|Convolutional   | 14x77x36  | filters=36, kernel=5x5, strides=2x2 |
|Activation      | 14x77x36  | RELU |
|Convolutional   | 5x37x48   | filters=48, kernel=5x5, strides=2x2 |
|Activation      | 5x37x48   | RELU |
|Convolutional   | 3x35x64   | filters=64, kernel=3x3, strides=1x1 |
|Activation      | 3x35x64   | RELU |
|Convolutional   | 1x33x64   | filters=64, kernel=3x3, strides=1x1 |
|Activation      | 1x33x64   | RELU |
|Flatten         | 2112      | |
|Fully connected | 100       | |
|Activation      | 100       | RELU |
|Fully connected | 50        | |
|Activation      | 50        | RELU |
|Fully connected | 10        | |
|Activation      | 10        | RELU |
|Output          | 1         | |

The network has 348219 parameters.

Here is a visualization of the architecture:

![text](img/model.png)

#### 3. Creation of the Training Set & Training Process

To capture good driving behavior, I first recorded four laps (two clockwise, and two counter-clockwise) on track one using center lane driving. Additionally, I captured driving through curve parts of the track one several times. Here is an example image of center lane driving:

![center](img/center.jpg)

I did not record the vehicle recovering from the left and right sides of the road back to center. Instead, I used images from the left and right cameras to train the model to recover. Here're example images taken by the left and right cameras.

![left](img/left.jpg) ![right](img/right.jpg)

Then I repeated this process on track two in order to get more data points.

To augment the data set, I also flipped images and angles thinking that this would help to combat overfitting and generalize the model. For example, here is an image that has then been flipped:

![flipped](img/center_flipped.jpg)

I used a keyboard for controlling the vehicle during capturing data on track one. This was good enough for the track on due to its simplicity. However this approach did not longer work for the track two. I had to use mouse to capture smooth steering, especially passing 180° turns.

After the collection process, I had 120834 data point. I then preprocessed this data by cropping each image (70 px from the top, and 25 px from the bottom), and normalizing it as `x = x / 255 - 0.5`.

I finally randomly shuffled the data set and put 20% of the data into a validation set. 

I used this training data for training the model. The validation set helped determine if the model was over or under fitting. The ideal number of epochs was 2 as evidenced by increasing the mean squared error during futher training. I used an adam optimizer so that manually training the learning rate wasn't necessary.

### Simulation

#### 1. The car is able to navigate correctly on test data

My final trained model is able to safely navigate the car on both track one and two. Here're example videos of the car driving on both tracks.

**Camera view**

<a href="http://www.youtube.com/watch?feature=player_embedded&v=4jov2hf09AU
" target="_blank"><img src="http://img.youtube.com/vi/4jov2hf09AU/0.jpg" 
alt="Track 1" width="240" height="180" border="10" /></a> <a href="http://www.youtube.com/watch?feature=player_embedded&v=5J8KsJsH8aY
" target="_blank"><img src="http://img.youtube.com/vi/5J8KsJsH8aY/0.jpg" 
alt="Track 2" width="240" height="180" border="10" /></a>

The video files are also available in the `video` directory.

**3rd-person view**

<a href="http://www.youtube.com/watch?feature=player_embedded&v=7FWQ3BUwcyg
" target="_blank"><img src="http://img.youtube.com/vi/7FWQ3BUwcyg/0.jpg" 
alt="Track 1" width="240" height="180" border="10" /></a> <a href="http://www.youtube.com/watch?feature=player_embedded&v=X4h_DA1XbwA
" target="_blank"><img src="http://img.youtube.com/vi/X4h_DA1XbwA/0.jpg" 
alt="Track 2" width="240" height="180" border="10" /></a>


history.md
## 2018-10-21 09:39pm

Changes / configuration:
- Added recording from the track 2 (CW + CCW), 2 laps for each direction using mouse (!!!) 
- I expect to see much stable behavior in the track 2

Result:
- The car passes the track 1 without any problems. It drives very well.
- The car passes the track 2 without any excedent. It doesn't drive perfect. But good enough to pass the track.

## 2018-10-21 02:18pm

Changes / configuration:
- Added recording from the track 2 (CW + CCW), 2 laps for each direction

Result:
- The car is able to pass most of the track 2 having speed of 7-10 Mph.
- However there're a few very curve turs the car cannot pass.


## 2018-10-21 11:37am 

Changes / configuration:
- Increased speed to 30 Mph

Result:
- The car still drives well! 
- But it cannot drive the track2

## 2018-10-21 11:17am 

Changes / configuration:
- Found and fix a bug in implementing the model (I used strides=(1,1) for the first 3 layers, while at NVidia they used strides=(2,2))

Result:
- The model size significantly decreased 1GB => 4Mb
- The card drives very well. I can try to increse speed in the next test.

## 2018-10-20 11:16pm

Changes / configuration:
- Increased the car's speed to 15

Result:
- The car is still able to pass the track, but now it is constantly moving from left to right and back.. ):
- Maybe it happens because of too high value of angle correction??

## 2018-10-20 06:25pm

Changes / configuration:
- Fixed a bug in the model (I forgot to use activations after conv and dense layers)
- Decreased the number of epochs to 2

Resuls:
- The car has finished the whole lap, it drives very well.
- The car eventually slows down. Don't know why.

## 2018-10-20 05:20pm

Changes / configuration:
- Added additional records for track 1 CW and CCW
- Added smooth recording for the most curve turns

Result:
- `Epoch 3/3 24076/24076 [==============================] - 522s - loss: 0.0532 - val_loss: 0.0596`
- The number of epoch is too much, 2 is more than okay.
- The car drives very bad. Going left, then right, then left. Cannot go strait.
- Maybe the model is overfitted?

## 2018-10-11 09:17pm

Changes / configuration:
- Crop 70 and 25 pixels from the rop and bottom respectively (instead of 60 and 20)
- Added files from clock-wise lap (`track1_cw`)

Result:
- `10968/10968 [==============================] - 130s 12ms/step - loss: 0.0128 - val_loss: 0.0115`
- The car is more centered now, but it still cannot pass the second curce turn.

## 2018-10-11 09:17pm

Changes / configuration:
- Crop 60 and 20 pixels from the top and bottom respectively

Result:
- `5568/5568 [==============================] - 81s 14ms/step - loss: 0.0094 - val_loss: 0.0378`
- The car passes the firt left turn and the bridge after it.
- However it failed to turn left at quite curve turn.

## 2018-10-11 09:17pm

Changes / configuration:
- Reduced the correction value to 0.1

Result:
- `5568/5568 [==============================] - 166s 30ms/step - loss: 0.0112 - val_loss: 0.0495`
- A car drives better, but it could not pass the first turn

## 2018-10-11 08:51pm

Changes / configuration:
- Using left and right camera images to train a model to correct its position
- Correction value is 0.2

Result:
- `5568/5568 [==============================] - 228s 41ms/step - loss: 0.0073 - val_loss: 0.0517`
- A car is steering right too much, need to tune correction value

## 2018-10-11 08:32pm

Changes / configuration:
- Added data augmentation by flipping original images

Result:
- `1856/1856 [==============================] - 73s 39ms/step - loss: 0.0185 - val_loss: 0.0552`
- The car is much more stable now

## 2018-10-11 08:28pm

Changes / configuration:
- Implemented LeNet-5
- Decreased the number of epochs to 3 as val_loss does not decrease

Result:
- `928/928 [==============================] - 37s 40ms/step - loss: 0.0799 - val_loss: 0.0520`
- A car is driving much better, but it is steeling left too much

## 2018-10-11 08:22pm

Changes / configuration:
- Normalized data as `x / 255.0 - 0.5`
- Simplest NN of one flatten layer
- 7 epochs
- Data from `track1_ccw`

Result:
- `928/928 [==============================] - 0s 431us/step - loss: 7062.8762 - val_loss: 5268.6844`

drive.py
import argparse
import base64
from datetime import datetime
import os
import shutil

import numpy as np
import socketio
import eventlet
import eventlet.wsgi
from PIL import Image
from flask import Flask
from io import BytesIO

from keras.models import load_model
import h5py
from keras import __version__ as keras_version

sio = socketio.Server()
app = Flask(__name__)
model = None
prev_image_array = None


class SimplePIController:
    def __init__(self, Kp, Ki):
        self.Kp = Kp
        self.Ki = Ki
        self.set_point = 0.
        self.error = 0.
        self.integral = 0.

    def set_desired(self, desired):
        self.set_point = desired

    def update(self, measurement):
        # proportional error
        self.error = self.set_point - measurement

        # integral error
        self.integral += self.error

        return self.Kp * self.error + self.Ki * self.integral


controller = SimplePIController(0.04, 0.002)
set_speed = 10
controller.set_desired(set_speed)


@sio.on('telemetry')
def telemetry(sid, data):
    if data:
        # The current steering angle of the car
        steering_angle = data["steering_angle"]
        # The current throttle of the car
        throttle = data["throttle"]
        # The current speed of the car
        speed = data["speed"]
        # The current image from the center camera of the car
        imgString = data["image"]
        image = Image.open(BytesIO(base64.b64decode(imgString)))
        image_array = np.asarray(image)
        steering_angle = float(model.predict(image_array[None, :, :, :], batch_size=1))

        throttle = controller.update(float(speed))

        print(steering_angle, throttle)
        send_control(steering_angle, throttle)

        # save frame
        if args.image_folder != '':
            timestamp = datetime.utcnow().strftime('%Y_%m_%d_%H_%M_%S_%f')[:-3]
            image_filename = os.path.join(args.image_folder, timestamp)
            image.save('{}.jpg'.format(image_filename))
    else:
        # NOTE: DON'T EDIT THIS.
        sio.emit('manual', data={}, skip_sid=True)


@sio.on('connect')
def connect(sid, environ):
    print("connect ", sid)
    send_control(0, 0)


def send_control(steering_angle, throttle):
    sio.emit(
        "steer",
        data={
            'steering_angle': steering_angle.__str__(),
            'throttle': throttle.__str__()
        },
        skip_sid=True)


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Remote Driving')
    parser.add_argument(
        'model',
        type=str,
        help='Path to model h5 file. Model should be on the same path.'
    )
    parser.add_argument(
        'image_folder',
        type=str,
        nargs='?',
        default='',
        help='Path to image folder. This is where the images from the run will be saved.'
    )
    args = parser.parse_args()

    # check that model Keras version is same as local Keras version
    f = h5py.File(args.model, mode='r')
    model_version = f.attrs.get('keras_version')
    keras_version = str(keras_version).encode('utf8')

    if model_version != keras_version:
        print('You are using Keras version ', keras_version,
              ', but the model was built using ', model_version)

    model = load_model(args.model)

    if args.image_folder != '':
        print("Creating image folder at {}".format(args.image_folder))
        if not os.path.exists(args.image_folder):
            os.makedirs(args.image_folder)
        else:
            shutil.rmtree(args.image_folder)
            os.makedirs(args.image_folder)
        print("RECORDING THIS RUN ...")
    else:
        print("NOT RECORDING THIS RUN ...")

    # wrap Flask application with engineio's middleware
    app = socketio.Middleware(sio, app)

    # deploy as an eventlet WSGI server
    eventlet.wsgi.server(eventlet.listen(('', 4567)), app)


model.py
from keras.models import Sequential
from keras.layers import Dense, Activation, Flatten, Lambda, Conv2D, MaxPooling2D, Cropping2D
from scipy import ndimage
import numpy as np
import csv
import cv2

"""
# Download data
wget -O data.zip https://www.dropbox.com/s/u7zy7414mu6vycs/carnd-behavioral-cloning-data2.zip?dl=0

# Unzip the archive
unzip data.zip -d ./data
"""

# Importing data.
# Each folder contains recorded images of a single run.
# 'track1_*' folders contain recorded images from runs in the track one.
# 'track2_*' folders contain recorded images from runs in the track two.
data_root_path = './data/'
folders = [
    'track1_ccw',
    'track1_ccw_2',
    'track1_ccw_turnA',
    'track1_ccw_turnB',
    'track1_cw',
    'track1_cw_2',
    'track1_cw_turnA',
    'track1_cw_turnB',
    'track2_cw_mouse',
    'track2_cw_mouse_2',
    'track2_ccw_mouse',
    'track2_ccw_mouse_2'
]

## Steering angle is in between -1 and 1 inclusive
## -1 implies the most left position (a car turns left)
##  1 implies the most right position (a car turns right)
steering_correction = 0.1

images = []
measurements = []

def load_image(data_path, source_image_path):
    """
    Load an image from filesystem as an numpy array.
    `data_path` - path to a folder containg images
    `source_image_path` - original path of the loading image
    """
    filename = source_image_path.split('/')[-1]
    current_path = data_path + '/IMG/' + filename
    return ndimage.imread(current_path)    

def flip_image(image, steering):
    """
    Flips the `image` horizontally.
    """
    image_flipped = np.fliplr(image)
    steering_flipped = -steering
    return (image_flipped, steering_flipped)

# Loading images from all specified folders
for folder in folders:
    lines = []
    data_path = data_root_path + folder
    print('Importing data from %s' % data_path)
    with open(data_path + '/driving_log.csv') as csvfile:
        reader = csv.reader(csvfile)
        for line in reader:
            lines.append(line)

    # For each line in the csv-file
    # 1. Load center, left and right images
    # 2. Calculate corrected steering angle for the left and right images
    # 3. Add flipped images
    for line in lines:
        center_image_path = line[0]
        left_image_path = line[1]
        right_image_path = line[2]

        image_center = load_image(data_path, center_image_path)
        image_left = load_image(data_path, left_image_path)
        image_right = load_image(data_path, right_image_path)
        
        steering_center = float(line[3])
        steering_left = steering_center + steering_correction
        steering_right = steering_center - steering_correction
        
        images.extend([image_center, image_left, image_right])
        measurements.extend([steering_center, steering_left, steering_right])

        image_center_flipped, steering_center_flipped = flip_image(image_center, steering_center)
        image_left_flipped, steering_left_flipped = flip_image(image_left, steering_left)
        image_right_flipped, steering_right_flipped = flip_image(image_right, steering_right)

        images.extend([image_center_flipped, image_left_flipped, image_right_flipped])
        measurements.extend([steering_center_flipped, steering_left_flipped, steering_right_flipped])

X_train = np.array(images)
y_train = np.array(measurements)

# Defining a model
model = Sequential()
model.add(Cropping2D(cropping=((70,25), (0,0)), input_shape=(160,320,3)))
model.add(Lambda(lambda x: x / 255.0 - 0.5))
model.add(Conv2D(filters=24, kernel_size=5, strides=(2,2), activation="relu"))
model.add(Conv2D(filters=36, kernel_size=5, strides=(2,2), activation="relu"))
model.add(Conv2D(filters=48, kernel_size=5, strides=(2,2), activation="relu"))
model.add(Conv2D(filters=64, kernel_size=3, activation="relu"))
model.add(Conv2D(filters=64, kernel_size=3, activation="relu"))
model.add(Flatten())
model.add(Dense(100, activation="relu"))
model.add(Dense(50, activation="relu"))
model.add(Dense(10, activation="relu"))
model.add(Dense(1))

model.summary()

# Compilation
model.compile(optimizer='adam', loss='mse')

# Training and saving the model
model.fit(X_train, y_train, validation_split=0.2, shuffle=True, epochs=2)
model.save('model.h5')
